trantor/net/inner/tlsprovider/BotanTLSProvider.cc

```diff
@@ -0,0 +1,390 @@
+        auto it =
+            std::find(cert_key_types.begin(), cert_key_types.end(), key_algo);
+        if (it == cert_key_types.end())
+            return {};
+        return {*cert_};
+    }
+
+    Botan::Private_Key *private_key_for(const Botan::X509_Certificate &cert,
+                                        const std::string &type,
+                                        const std::string &context) override
+    {
+        (void)cert;
+        (void)type;
+        (void)context;
+        return key_;
+    }
+    Botan::Certificate_Store *certStore_ = nullptr;
+    Botan::X509_Certificate *cert_ = nullptr;
+    Botan::Private_Key *key_ = nullptr;
+};
+
+struct BotanCertificate : public Certificate
+{
+    BotanCertificate(const Botan::X509_Certificate &cert) : cert_(cert)
+    {
+    }
+
+    virtual std::string sha1Fingerprint() const override
+    {
+        return cert_.fingerprint("SHA-1");
+    }
+
+    virtual std::string sha256Fingerprint() const override
+    {
+        return cert_.fingerprint("SHA-256");
+    }
+
+    virtual std::string pem() const override
+    {
+        return cert_.PEM_encode();
+    }
+    Botan::X509_Certificate cert_;
+};
+
+namespace trantor
+{
+struct SSLContext
+{
+    std::unique_ptr<Botan::Private_Key> key;
+    std::unique_ptr<Botan::X509_Certificate> cert;
+    std::unique_ptr<Botan::Certificate_Store> certStore;
+    bool isServer = false;
+    bool requireClientCert = false;
+};
+}  // namespace trantor
+
+class TrantorPolicy : public Botan::TLS::Policy
+{
+    virtual bool require_cert_revocation_info() const override
+    {
+        return false;
+    }
+
+    virtual bool require_client_certificate_authentication() const override
+    {
+        return requireClientCert_;
+    }
+
+  public:
+    bool requireClientCert_ = false;
+};
+
+struct BotanTLSProvider : public TLSProvider,
+                          public NonCopyable,
+                          public Botan::TLS::Callbacks
+{
+  public:
+    BotanTLSProvider(TcpConnection *conn,
+                     TLSPolicyPtr policy,
+                     SSLContextPtr ctx)
+        : TLSProvider(conn, std::move(policy), std::move(ctx))
+    {
+    }
+
+    virtual void recvData(MsgBuffer *buffer) override
+    {
+        LOG_TRACE << "Low level connection received " << buffer->readableBytes()
+                  << " bytes.";
+        try
+        {
+            assert(channel_ != nullptr);
+            channel_->received_data((const uint8_t *)buffer->peek(),
+                                    buffer->readableBytes());
+        }
+        catch (const Botan::TLS::TLS_Exception &e)
+        {
+            LOG_ERROR << "Unexpected TLS Exception: " << e.what();
+            conn_->shutdown();
+
+            if (tlsConnected_ == false)
+                handleSSLError(SSLError::kSSLHandshakeError);
+            else
+                handleSSLError(SSLError::kSSLProtocolError);
```

Comment:
Maybe inspect the `TLS_Exception` more thoroughly to provide better error categories to your `handleSSLError()`. As described below (in `tls_session_established`):

```suggestion
            if (tlsConnected_ == false) {
                if(e.type() == Botan::Alert::BAD_CERTIFICATE)
                    handleSSLError(SSLError::kSSLInvalidCertificate);
                else
                    handleSSLError(SSLError::kSSLHandshakeError);
            }
            else
                handleSSLError(SSLError::kSSLProtocolError);
```